File: src/main/java/com/warren/bt/Parallel.java
package com.warren.bt;

public class Parallel extends Composite {

  public enum Policy {
    REQUIRE_ONE, // Succeed when one child succeeds
    REQUIRE_ALL  // Succeed when all children succeed
  }

  private final Policy successPolicy;
  private final Policy failurePolicy;

  public Parallel(Policy successPolicy, Policy failurePolicy) {
    this.successPolicy = successPolicy;
    this.failurePolicy = failurePolicy;
  }

  public Parallel() {
    this(Policy.REQUIRE_ALL, Policy.REQUIRE_ONE);
  }

  @Override
  public Status tick() {
    if (!hasChildren()) {
      return Status.SUCCESS;
    }

    int successCount = 0;
    int failureCount = 0;
    int runningCount = 0;

    for (Node child : children) {
      Status status = child.tick();

      switch (status) {
        case SUCCESS:
          successCount++;
          break;
        case FAILURE:
          failureCount++;
          break;
        case RUNNING:
          runningCount++;
          break;
      }
    }

    // Failure conditions
    if (failurePolicy == Policy.REQUIRE_ONE && failureCount > 0) {
      reset();
      return Status.FAILURE;
    }

    if (failurePolicy == Policy.REQUIRE_ALL && failureCount == children.size()) {
      reset();
      return Status.FAILURE;
    }

    // Success conditions
    if (successPolicy == Policy.REQUIRE_ONE && successCount > 0) {
      reset();
      return Status.SUCCESS;
    }

    if (successPolicy == Policy.REQUIRE_ALL && successCount == children.size()) {
      reset();
      return Status.SUCCESS;
    }

    // If none succeeded or failed fully, still active
    return Status.RUNNING;
  }
}

File: src/main/kotlin/com/warren/bt/Parallel.kt
package com.warren.bt

class Parallel(
        private val successPolicy: Policy = Policy.REQUIRE_ALL,
        private val failurePolicy: Policy = Policy.REQUIRE_ONE
) : Composite("Parallel") {

  enum class Policy {
    REQUIRE_ALL,
    REQUIRE_ONE
  }

  override fun tick(): Status {
    if (children.isEmpty()) return Status.SUCCESS

    var successCount = 0
    var failureCount = 0
    var runningFound = false

    for (c in children) {
      when (c.tick()) {
        Status.SUCCESS -> successCount++
        Status.FAILURE -> failureCount++
        Status.RUNNING -> runningFound = true
      }
    }

    val all = children.size

    val success =
            when (successPolicy) {
              Policy.REQUIRE_ALL -> successCount == all
              Policy.REQUIRE_ONE -> successCount > 0
            }
    val failure =
            when (failurePolicy) {
              Policy.REQUIRE_ALL -> failureCount == all
              Policy.REQUIRE_ONE -> failureCount > 0
            }

    return when {
      success -> Status.SUCCESS
      failure -> Status.FAILURE
      else -> Status.RUNNING
    }
  }
}
